Add declarativeNetRequest (#22644)
mdn/content: files/en-us/mozilla/add-ons/webextensions/api/declarativenetrequest/index.md

- enable static rulesets up to at least the value of {{WebExtAPIRef("declarativeNetRequest.GUARANTEED_MINIMUM_STATIC_RULES","GUARANTEED_MINIMUM_STATIC_RULES")}}, and the number of enabled static rulesets must not exceed the value of [{{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_ENABLED_STATIC_RULESETS","MAX_NUMBER_OF_ENABLED_STATIC_RULESETS")}}. In addition, the number of rules in enabled static rulesets for all extensions must not exceed the global limit. Extensions shouldn't depend on the global limit having a specific value and should instead use {{WebExtAPIRef("declarativeNetRequest.getAvailableStaticRuleCount","getAvailableStaticRuleCount")}} to find the number of additional rules they can enable.

### Dynamic and session-scoped rules

The number of dynamic and session-scoped rules an extension can add is limited to the value of {{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES","MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES")}}.

## Matching precedents

> Before a request is sent, each extension is queried for an action to take. In order of precedence, these actions are:
> 
> 1. Actions that block requests of type `block`.
> 2. Actions that redirect requests of type `redirect` or `upgradeScheme`.
> 3. Actions that allow requests of type `allow` or `allowAllRequests`.
> 
> If more than one extension returns an action, the extension whose action has the highest precedence gets priority. If more than one extension returns an action with the same priority, the most recently installed extension gets priority.
> 
> When an extension is queried for how to handle a request, the matching rule with the highest `priority` is returned. If more than one matching rule has the highest `priority`, the tie is broken based on the action type, in this order of precedence:
> 
> 1. `allow`
> 2. `allowAllRequests`
> 3. `block`
> 4. `upgradeScheme`
> 5. `redirect`

Author: @Rob--W Is this accurate? However, could you still end up with multiple rules for "redirect" or "modifyHeaders" actions? The latter is, I assume, covered by the note - but what about "redirect"
```suggestion
When the browser evaluates how to handle requests, it chooses among all the rules that have a condition that matches the request as follows:

1. the ruleset the rule belongs to, in this order of precedence:
  1. session
  2. dynamic
  3. static
  If this doesn't result in one rule to evaluate:
2. the order of the rule in the ruleset, determined as the lowest value rule ID.
   If this doesn't result in one rule to evaluate:
3. the rule priority, where 1 is the highest priority (and rules default to 1 where priority is not set)
   If this doesn't result in one rule to evaluate:
4. the rule action, in the following order of precedence:
  1. "allow" which means any other remaining rules are ignored.
  2. "allowAllRequests" (for main_frame and sub_frame resourceTypes only) has the same effect as allow but also applies to future subresource loads in the document (including descendant frames) generated from the request.
  3. "block" cancels the request.
  4. "upgradeScheme" upgrades the scheme of the request.
  5. "redirect" redirects the request.
  6. "modifyHeaders" rewrites request and response headers. Note that a header may be modified only once, with one exception. The "append" operation can be executed again if the previous operation is "set" or "append"
```

Reviewer: The above is almost accurate, at least for rule matching within an extension.

"3. the rule priority," should be at the top. `rule.priority` takes precedence over everything else. See this part of the explanation: https://searchfox.org/mozilla-central/rev/9dfda5ccb0fc42d7666a54b1caf1af6525e49694/toolkit/components/extensions/ExtensionDNR.sys.mjs#33-36

 The matching behavior within an extension is the most important part that an extension can control, but for completeness we also need to document what happens when multiple extensions provide an action. It's somewhat confusing that Chrome's docs start with how conflicts between extensions are resolved, when an extension author should primarily be concerned with their own DNR rules. In my previously linked code comment, this is the brief part that explains the conflict resolution between extensions: https://searchfox.org/mozilla-central/rev/9dfda5ccb0fc42d7666a54b1caf1af6525e49694/toolkit/components/extensions/ExtensionDNR.sys.mjs#63-69

> However, could you still end up with multiple rules for "redirect" or "modifyHeaders" actions? The latter is, I assume, covered by the note - but what about "redirect"

It is always possible to end up with multiple potential matches, but if the request ends up being canceled/replaced, the algorithm is guaranteed to only choose one rule. Even if we have two distinct actions for two otherwise identical rules (ruleset, priority, condition), the API requires a unique ID to be specified. And because of that, we can always choose the highest-precedence rule. Here is another explanation of exactly the same topic: https://github.com/w3c/webextensions/issues/280#issuecomment-1262607883

I hope that you're able to use this to generate readable documentation :)

Author: @Rob--W This should incorporate your feedback.
```suggestion
When the browser evaluates how to handle requests, it checks each extension's rules that have a condition that matches the request and chooses the one to consider applying as follows:
1. the rule priority, where 1 is the highest priority (and rules default to 1 where priority is not set)
   If this doesn't result in one rule to apply:
2. the ruleset the rule belongs to, in this order of precedence:
  1. session
  2. dynamic
  3. static
  If this doesn't result in one rule to apply:
3. the order of the rule in the ruleset, determined as the lowest value rule ID.
   If this doesn't result in one rule to apply:
4. the rule action, in the following order of precedence:
  1. "allow" which means any other remaining rules are ignored.
  2. "allowAllRequests" (for main_frame and sub_frame resourceTypes only) has the same effect as allow but also applies to future subresource loads in the document (including descendant frames) generated from the request.
  3. "block" cancels the request.
  4. "upgradeScheme" upgrades the scheme of the request.
  5. "redirect" redirects the request.
  6. "modifyHeaders" rewrites request and response headers. Note that a header may be modified only once, with one exception. The "append" operation can be executed again if the previous operation is "set" or "append".

If only one extension provides a rule for the request, that rule is applied. However, where more than one extension has a matching rule, the browser chooses the one to apply in this order of precedence:
1. "block"
2. "redirect" and "upgradeScheme"
3. "allow" and "allowAllRequests"
```

Reviewer: I have modified your edit as follows:

- Move 4 to between 1 and 2 (I accidentally overlooked that before).
- Remove elaborate description of "modifyHeaders" because that is not relevant for rule matching.
- Appended a brief paragraph on when `modifyHeaders` applied plus a reference to the ModifyHeaderInfo docs.

Feel free to edit this suggested edit and commit the final suggestion.

```suggestion
When the browser evaluates how to handle requests, it checks each extension's rules that have a condition that matches the request and chooses the one to consider applying as follows:
1. the rule priority, where 1 is the highest priority (and rules default to 1 where priority is not set)
   If this doesn't result in one rule to apply:
2. the rule action, in the following order of precedence:
  1. "allow" which means any other remaining rules are ignored.
  2. "allowAllRequests" (for main_frame and sub_frame resourceTypes only) has the same effect as allow but also applies to future subresource loads in the document (including descendant frames) generated from the request.
  3. "block" cancels the request.
  4. "upgradeScheme" upgrades the scheme of the request.
  5. "redirect" redirects the request.
  6. "modifyHeaders" rewrites request and response headers.
   If this doesn't result in one rule to apply:
3. the ruleset the rule belongs to, in this order of precedence:
  1. session
  2. dynamic
  3. static
  If this doesn't result in one rule to apply:
4. the order of the rule in the ruleset, determined as the lowest value rule ID.

If only one extension provides a rule for the request, that rule is applied. However, where more than one extension has a matching rule, the browser chooses the one to apply in this order of precedence:
1. "block"
2. "redirect" and "upgradeScheme"
3. "allow" and "allowAllRequests"

If the request was not blocked or redirected, the matching `modifyHeaders` actions are applied, as documented in {{WebExtAPIRef("declarativeNetRequest.ModifyHeaderInfo")}}.
```

When an extension is queried for how to handle a request, the matching rule with the highest `priority` is returned. If more than one matching rule has the highest `priority`, the tie is broken based on the action type, in this order of precedence:

1. `allow`
2. `allowAllRequests`
3. `block`
4. `upgradeScheme`
5. `redirect`

If the request is not blocked or redirected, the matching `modifyHeaders` rules are evaluated, with the most recently installed extensions getting priority. Within each extension, all `modifyHeaders` rules with a priority lower than matching `allow` or `allowAllRequests` rules are ignored.

If multiple `modifyHeaders` rules specify the same header, the resulting modification for the header is determined based on the priority of each rule and the operations specified:

- If a rule has been appended to a header, then lower-priority rules can only append to that header. `set` and `remove` operations are not permitted.
- If a rule has set a header, lower priority rules cannot further modify the header, except for `append` rules from the same extension.
- If a rule has removed a header, lower priority rules cannot further modify the header.

## Testing

> 

Author: ```suggestion
```

- enable static rulesets up to at least the value of {{WebExtAPIRef("declarativeNetRequest.GUARANTEED_MINIMUM_STATIC_RULES","GUARANTEED_MINIMUM_STATIC_RULES")}}, and the number of enabled static rulesets must not exceed the value of [{{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_ENABLED_STATIC_RULESETS","MAX_NUMBER_OF_ENABLED_STATIC_RULESETS")}}. In addition, the number of rules in enabled static rulesets for all extensions must not exceed the global limit. Extensions shouldn't depend on the global limit having a specific value and should instead use {{WebExtAPIRef("declarativeNetRequest.getAvailableStaticRuleCount","getAvailableStaticRuleCount")}} to find the number of additional rules they can enable.

### Dynamic and session-scoped rules

The number of dynamic and session-scoped rules an extension can add is limited to the value of {{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES","MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES")}}.

## Matching precedents

When the browser evaluates how to handle requests, it checks each extension's rules that have a condition that matches the request and chooses the one to consider applying as follows:

> 1. the rule priority, where 1 is the highest priority (and rules default to 1 where priority is not set)

Reviewer: ```suggestion
1. the rule priority, where 1 is the lowest priority (and rules default to 1 where priority is not set)
```

- (Chrome only:) Unlike the webRequest API, any images or iframes blocked using the declarativeNetRequest API are automatically collapsed in the DOM.
- While deciding whether a request is to be blocked or redirected, the declarativeNetRequest API is given priority over the webRequest API because it allows for synchronous interception. Similarly, any headers removed through declarativeNetRequest API are not made visible to web request extensions.
- The webRequest API is more flexible than the declarativeNetRequest API because it allows extensions to evaluate a request programmatically.

## Types

- {{WebExtAPIRef("declarativeNetRequest.MatchedRule")}}
  - : Details of a matched rule.
- {{WebExtAPIRef("declarativeNetRequest.ModifyHeaderInfo")}}
  - : The request or response headers to modify for the request.
- {{WebExtAPIRef("declarativeNetRequest.Redirect")}}
  - : Details of how the redirect should be performed. Only valid for redirect rules.
- {{WebExtAPIRef("declarativeNetRequest.ResourceType")}}
  - : The resource type of a request.
- {{WebExtAPIRef("declarativeNetRequest.Rule")}}

>   - : An object containing details of a dynamic rule.

Reviewer: ```suggestion
  - : An object containing details of a rule.
```

Not just dynamic, also session rules. And also describes the format of a static rule.

## Properties

- {{WebExtAPIRef("declarativeNetRequest.DYNAMIC_RULESET_ID")}}
  - : Ruleset ID for the dynamic rules added by the extension.
- {{WebExtAPIRef("declarativeNetRequest.GETMATCHEDRULES_QUOTA_INTERVAL")}}
  - : The time interval within which {{WebExtAPIRef("declarativeNetRequest.MAX_GETMATCHEDRULES_CALLS_PER_INTERVAL")}} {{WebExtAPIRef("declarativeNetRequest.getMatchedRules")}} calls can be made.
- {{WebExtAPIRef("declarativeNetRequest.GUARANTEED_MINIMUM_STATIC_RULES")}}
  - : The minimum number of static rules guaranteed to an extension across its enabled static rulesets.
- {{WebExtAPIRef("declarativeNetRequest.MAX_GETMATCHEDRULES_CALLS_PER_INTERVAL")}}
  - : The number of times {{WebExtAPIRef("declarativeNetRequest.getMatchedRules")}} can be called within a period of {{WebExtAPIRef("declarativeNetRequest.GETMATCHEDRULES_QUOTA_INTERVAL")}}.
- {{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES")}}
  - : The maximum number of combined dynamic and session scoped rules an extension can add.
- {{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_ENABLED_STATIC_RULESETS")}}

>   - : The maximum number of static `Rulesets` an extension can enable.

Reviewer: ```suggestion
  - : The maximum number of static rulesets an extension can enable.
```

- {{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_REGEX_RULES")}}
  - : The maximum number of regular expression rules that an extension can add.
- {{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_STATIC_RULESETS")}}

>   - : The maximum number of static `Rulesets` an extension can specify as part of the [`"rule_resources"`](/docs/Mozilla/Add-ons/WebExtensions/manifest.json/rule_resources) manifest key.

Reviewer: ```suggestion
  - : The maximum number of static rulesets an extension can specify as part of the [`declarative_net_request.rule_resources`](/docs/Mozilla/Add-ons/WebExtensions/manifest.json/declarative_net_request) manifest key.
```

## Permissions

To use this API, an extension must request the `"declarativeNetRequest"` or `"declarativeNetRequestWithHostAccess"` [permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) in its [`manifest.json`](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json) file.

The `"declarativeNetRequest"` permission allows extensions to block and upgrade requests without any [host permissions](/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions). Host permissions are required if the extension wants to redirect a request or modify headers on a request. The `"declarativeNetRequestWithHostAccess"` permission requires host permissions to the request URL and initiator to act on a request.

The `"declarativeNetRequestFeedback"` permission is required to use {{WebExtAPIRef("declarativeNetRequest.getmatchedrules","getmatchedrules")}} and {{WebExtAPIRef("declarativeNetRequest.onRuleMatchedDebug","onRuleMatchedDebug")}} as they return information on declarative rules matched.

## Rules

> The declarative rules are defined by four fields:

Reviewer: The content here is almost the same as the `Rule` docs. Can we consolidate the two somehow? Or at least interlink them?

Author: I think it makes sense to have this content here. Otherwise, the overview discussion is going to be disjointed. I don't see the need to link through to the documentation for the `Rule` type, as that follows through from the links to the methods in the section on rulesets. However, I have put a link back to here from the `Rule` type documentation.

- enable static rulesets up to at least the value of {{WebExtAPIRef("declarativeNetRequest.GUARANTEED_MINIMUM_STATIC_RULES","GUARANTEED_MINIMUM_STATIC_RULES")}}, and the number of enabled static rulesets must not exceed the value of [{{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_ENABLED_STATIC_RULESETS","MAX_NUMBER_OF_ENABLED_STATIC_RULESETS")}}. In addition, the number of rules in enabled static rulesets for all extensions must not exceed the global limit. Extensions shouldn't depend on the global limit having a specific value and should instead use {{WebExtAPIRef("declarativeNetRequest.getAvailableStaticRuleCount","getAvailableStaticRuleCount")}} to find the number of additional rules they can enable.

### Dynamic and session-scoped rules

The number of dynamic and session-scoped rules an extension can add is limited to the value of {{WebExtAPIRef("declarativeNetRequest.MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES","MAX_NUMBER_OF_DYNAMIC_AND_SESSION_RULES")}}.

> ## Matching precedents

Reviewer: The current documentation as copied from Chrome is unclear. There is ambiguity over what actions have the highest precedence. There are undocumented orders of precedence (ruleset and rule ID). The "allow" and "allowAllRequests" actions are poorly documented. Their name is based on DNR's initial design, which was to either block or allow a request. "redirect" was added later, at which point "allow" basically meant "ignore other actions (such as block, redirect, upgradeScheme)".

Here is an attempt to have complete yet succinct documentation of the matching behavior: https://searchfox.org/mozilla-central/rev/270f0ed2146821e239232728adac41a3de41131b/toolkit/components/extensions/ExtensionDNR.sys.mjs#10-70
Could you try to use that as the basis for this section of the documentation instead of Chrome's?

Before a request is sent, each extension is queried for an action to take. In order of precedence, these actions are:

1. Actions that block requests of type `block`.
2. Actions that redirect requests of type `redirect` or `upgradeScheme`.
3. Actions that allow requests of type `allow` or `allowAllRequests`.

If more than one extension returns an action, the extension whose action has the highest precedence gets priority. If more than one extension returns an action with the same priority, the most recently installed extension gets priority.

When an extension is queried for how to handle a request, the matching rule with the highest `priority` is returned. If more than one matching rule has the highest `priority`, the tie is broken based on the action type, in this order of precedence:

1. `allow`
2. `allowAllRequests`
3. `block`
4. `upgradeScheme`
5. `redirect`

If the request is not blocked or redirected, the matching `modifyHeaders` rules are evaluated, with the most recently installed extensions getting priority. Within each extension, all `modifyHeaders` rules with a priority lower than matching `allow` or `allowAllRequests` rules are ignored.

> If multiple `modifyHeaders` rules specify the same header, the resulting modification for the header is determined based on the priority of each rule and the operations specified:

Reviewer: This whole section about the applicable header modifications would better be merged with the modifyheaderinfo/index.md file. I have already provided a suggested edit; the three bullet points below can be summarized to `A header may be modified only once, with one exception. The "append" operation can be executed again if the previous operation is "set" or "append".` (this is what I suggested in modifyheaderinfo/index.md)

Author: @Rob--W by whole section do you mean:
```
If the request is not blocked or redirected, the matching `modifyHeaders` rules are evaluated, with the most recently installed extensions getting priority. Within each extension, all `modifyHeaders` rules with a priority lower than matching `allow` or `allowAllRequests` rules are ignored.

If multiple `modifyHeaders` rules specify the same header, the resulting modification for the header is determined based on the priority of each rule and the operations specified:

- If a rule has been appended to a header, then lower-priority rules can only append to that header. `set` and `remove` operations are not permitted.
- If a rule has set a header, lower priority rules cannot further modify the header, except for `append` rules from the same extension.
- If a rule has removed a header, lower priority rules cannot further modify the header.
```

- If a rule has been appended to a header, then lower-priority rules can only append to that header. `set` and `remove` operations are not permitted.
- If a rule has set a header, lower priority rules cannot further modify the header, except for `append` rules from the same extension.
- If a rule has removed a header, lower priority rules cannot further modify the header.

## Testing

{{WebExtAPIRef("declarativeNetRequest.getmatchedrules","getmatchedrules")}} and {{WebExtAPIRef("declarativeNetRequest.onRuleMatchedDebug","onRuleMatchedDebug")}} are available to assist with testing rules and rulesets. These APIs require the `"declarativeNetRequestFeedback"` [permissions](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions). In addition:

> - in Chrome, these APIs are only available to unpacked extensions.

Reviewer: ```suggestion
- in Chrome, these APIs are only available to unpacked extensions.
- in Firefox, these APIs are only available after setting the `extensions.dnr.feedback` preference to `true`. This preference can be set via `about:config` or with the [`--pref` flag of the `web-ext` CLI tool](https://extensionworkshop.com/documentation/develop/web-ext-command-reference/#pref).
```

(This would resolve https://github.com/mdn/content/pull/22644#discussion_r1042467218 )

Author: ```suggestion
- in Chrome, these APIs are only available to unpacked extensions.
- in Firefox, these APIs are only available after setting the `extensions.dnr.feedback` preference to `true`. Set this preference using `about:config` or the [`--pref` flag of the `web-ext` CLI tool](https://extensionworkshop.com/documentation/develop/web-ext-command-reference/#pref).
```

Reviewer: Your suggestion looks good. When we do end up changing the implementation, we can update the documentation again. E.g. if https://bugzilla.mozilla.org/show_bug.cgi?id=1811947#c2 is implemented, it would become:

```suggestion
- in Chrome, these APIs are only available to unpacked extensions.
- in Firefox, these APIs are only available to temporarily loaded extensions.
```

Reviewer: @rebloor let's also include the top-level sugggested edit.

The implementation settled on continuing to require the preference.

Author: @Rob--W thanks for the confirmation, done

## Comparison with the webRequest API

- The declarativeNetRequest API evaluates network requests in the browser itself. This makes it more performant than the webRequest API, where each network request is evaluated in JavaScript in the extension process.

> - Because the requests are not intercepted by the extension process, declarativeNetRequest removes the need for extensions to have a background page, resulting in less memory consumption.

Reviewer: ```suggestion
- Because the requests are not intercepted by the extension process, declarativeNetRequest removes the need for extensions to have a background page.
```

- Unlike the webRequest API, blocking or upgrading requests using the declarativeNetRequest API requires no host permissions when used with the `declarativeNetRequest` permission.
- The declarativeNetRequest API provides better privacy to users because extensions do not read the network requests made on the user's behalf.

> - Unlike the webRequest API, any images or iframes blocked using the declarativeNetRequest API are automatically collapsed in the DOM.

Reviewer: ```suggestion
- (Chrome only:) Unlike the webRequest API, any images or iframes blocked using the declarativeNetRequest API are automatically collapsed in the DOM.
```

- While deciding whether a request is to be blocked or redirected, the declarativeNetRequest API is given priority over the webRequest API because it allows for synchronous interception. Similarly, any headers removed through declarativeNetRequest API are not made visible to web request extensions.
- The webRequest API is more flexible than the declarativeNetRequest API because it allows extensions to evaluate a request programmatically.

> ## Example

Reviewer: Let's delete this whole Example section. It is not a good representation of how the API can/should be used. I'll author a new example eventually and submit a PR to this repo and https://github.com/mdn/webextensions-examples when the time comes.

  },
  {
    "id": 4,
    "priority": 1,
    "action": { "type": "redirect", "redirect": { "url": "https://example.com" } },
    "condition": { "urlFilter": "google.com", "resourceTypes": ["main_frame"] }
  },
  {
    "id": 5,
    "priority": 1,
    "action": { "type": "redirect", "redirect": { "extensionPath": "/a.jpg" } },
    "condition": { "urlFilter": "abcd.com", "resourceTypes": ["main_frame"] }
  },
  {
    "id": 6,
    "priority": 1,
    "action": {
      "type": "redirect",
      "redirect": {
        "transform": { "scheme": "https", "host": "new.example.com" }
      }
    },
    "condition": { "urlFilter": "||example.com", "resourceTypes": ["main_frame"] }
  },
  {
    "id": 7,
    "priority": 1,
    "action": {
      "type": "redirect",
      "redirect": {
        "regexSubstitution": "https://\\1.xyz.com/"
      }
    },
    "condition": {
      "regexFilter": "^https://www\\.(abc|def)\\.xyz\\.com/",
      "resourceTypes": [
        "main_frame"
      ]
    }
  },
  {
    "id" : 8,
    "priority": 2,
    "action" : {
      "type" : "allowAllRequests"
    },
    "condition" : {
      "urlFilter" : "||b.com/path",
      "resourceTypes" : ["sub_frame"]

>     }
>   },
>   {
>     "id" : 9,
>     "priority": 1,
>     "action" : {
>       "type" : "block"
>     },
>     "condition" : {
>       "urlFilter" : "script.js",
>       "resourceTypes" : ["script"]
>     }
>   },
>   {
>     "id": 10,
>     "priority": 2,
>     "action": {
>       "type": "modifyHeaders",
>       "responseHeaders": [
>         { "header": "h1", "operation": "remove" },
>         { "header": "h2", "operation": "set", "value": "v2" },
>         { "header": "h3", "operation": "append", "value": "v3" }
>       ]
>     },
>     "condition": { "urlFilter": "headers.com/123", "resourceTypes": ["main_frame"] }
>   },
>   {
>     "id": 11,
>     "priority": 1,
>     "action": {
>       "type": "modifyHeaders",
>       "responseHeaders": [
>         { "header": "h1", "operation": "set", "value": "v4" },
>         { "header": "h2", "operation": "append", "value": "v5" },
>         { "header": "h3", "operation": "append", "value": "v6" }
>       ]
>     },
>     "condition": { "urlFilter": "headers.com/12345", "resourceTypes": ["main_frame"] }
>   },
> ]
> ```
> 
> To illustrate how this rule set affects requests, consider:
> 
> - navigation to `"https://google.com"`. Rules with ID (1) and (4) match. The request iw blocked because blocking rules have higher priority than redirect rules when the `"priority"` is the same.
> - navigation to `"https://google.com/1234"`. Rules with ID (1), (2), and (4) match. Because the request has a matching `allow` rule and no higher priority rules, the request is not blocked nor redirected and continues to `"https://google.com/1234"`.
> - navigation to `"https://google.com/12345"` Rules with ID (1), (2), (3), and (4) match. The request is blocked because rule (3) has the highest priority, overriding all other matching rules.
> - navigation to `"https://abcd.com"`. The rule with ID (5) matches. Since rule (5) specifies an extension path, the request is redirected to `"chrome-extension://EXTENSION_ID/a.jpg"`.
> - navigation to `"http://example.com/path"`. The rule with ID (6) matches. Since rule (6) specifies a URL transform, the request is redirected to `"https://new.example.com/path"`.
> - a navigation to `"https://www.abc.xyz.com/path"`. The rule with ID (7) matches. The request is redirected to `"https://abc.xyz.com/path"`.
> - this request hierarchy:
>   - <https://a.com/path> (main-frame request)
>     - <https://b.com/path> (sub-frame request, matches the rule with ID (8))
>       - <https://c.com/path> (sub-frame request, matches the rule with ID (8))
>         - <https://c.com/script.js> (script request, matches the rules with IDs (8, 9) but (8) has higher priority)
>       - <https://b.com/script.js> (script request, matches the rules with IDs (8, 9) but (8) has higher priority)
>     - <https://d.com/path> (sub-frame request)
>       - <https://d.com/script.js> (script request, matches the rule with ID (9)) All requests in green will be allow-listed due to rule with ID (8) and not be evaluated by the extensions' ruleset. Requests in red are blocked due to rule with ID (9).
> - navigation to `"https://headers.com/12345"` with response headers `{ "h1": "initial_1", "h2": "initial_2" }`. Rules with ID (10) and (11) match. The request has its response headers modified to `{ "h2": "v2", "h2": "v5", "h3": "v3", "h3": "v6" }`. Header `h1` is removed by (10), `h2` is set by (10) then appended by (11), and `h3` is appended by (10) and (11).
> 

Author: ```suggestion
```

## Types

- {{WebExtAPIRef("declarativeNetRequest.MatchedRule")}}
  - : Details of a matched rule.
- {{WebExtAPIRef("declarativeNetRequest.ModifyHeaderInfo")}}

>   - : The request headers to modify for the request.

Reviewer: ```suggestion
  - : The request or response headers to modify for the request.
```

- {{WebExtAPIRef("declarativeNetRequest.Redirect")}}
  - : Details of how the redirect should be performed. Only valid for redirect rules.
- {{WebExtAPIRef("declarativeNetRequest.ResourceType")}}
  - : The resource type of a request.
- {{WebExtAPIRef("declarativeNetRequest.Rule")}}
  - : An object containing details of a dynamic rule.
- {{WebExtAPIRef("declarativeNetRequest.RuleAction")}}
  - : An object defining the action to take if a rule is matched.
- {{WebExtAPIRef("declarativeNetRequest.RuleCondition")}}
  - : An object defining the condition under which a rule is triggered.

> - {{WebExtAPIRef("declarativeNetRequest.Ruleset")}}
>   - : An object containing details of a static ruleset.

Reviewer: As mentioned in that article, it's already covered by the manifest docs.
```suggestion
```

- {{WebExtAPIRef("declarativeNetRequest.URLTransform")}}

>   - : An object containing details of a URL transformation to perform.

Reviewer: ```suggestion
  - : An object containing details of a URL transformation to perform for a redirect action.
```
